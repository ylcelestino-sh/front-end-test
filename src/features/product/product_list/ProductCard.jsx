import React from "react";
import PropTypes from "prop-types";
import {
  Box,
  Center,
  Heading,
  Text,
  Stack,
  useColorModeValue,
} from "@chakra-ui/react";
import { useNavigate } from "react-router-dom";
import LazyImage from "../../../components/LazyImage";

const ProductCard = ({ image, brand, model, price, id }) => {
  const navigate = useNavigate();
  return (
    <Center py={12} pb={2} data-testid="productCard">
      <Box
        role={"group"}
        p={2}
        maxW={"230px"}
        w={"full"}
        bg={useColorModeValue("white", "gray.800")}
        boxShadow={"2xl"}
        rounded={"lg"}
        pos={"relative"}
        zIndex={1}
        as="button"
        data-testid="productCardBox"
        onClick={() => navigate(`/product/${id}`)}
      >
        <Box
          rounded={"lg"}
          mt={-12}
          pos={"relative"}
          height={"130px"}
          _after={{
            transition: "all .5s ease",
            content: '""',
            w: "full",
            h: "full",
            pos: "absolute",
            top: 5,
            left: 0,
            filter: "blur(15px)",
            zIndex: -1,
          }}
          _groupHover={{
            _after: {
              filter: "blur(20px)",
            },
          }}
        >
          <LazyImage
            alt={model}
            src={image}
            rounded={"lg"}
            height={150}
            width={282}
            objectFit={"contain"}
          />
        </Box>
        <Stack pt={10}>
          <Heading as="h1" size="md">
            {model}
          </Heading>
          <Heading as="h2" size="md">
            ${price}
          </Heading>
          <Stack direction={"column"}>
            <Text fontWeight={100} fontSize={"sm"}>
              brand: {brand}
            </Text>
          </Stack>
        </Stack>
      </Box>
    </Center>
  );
};

ProductCard.propTypes = {
  image: PropTypes.string,
  brand: PropTypes.string,
  model: PropTypes.string,
  price: PropTypes.string,
  id: PropTypes.string,
};
export default ProductCard;
