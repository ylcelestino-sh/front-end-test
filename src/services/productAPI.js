import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";

export const productAPI = createApi({
  reducerPath: "products",
  baseQuery: fetchBaseQuery({
    baseUrl: "https://front-test-api.herokuapp.com/api/",
  }),
  tagTypes: ['Product'],
  keepUnusedDataFor: 3600, // it's delete unused data when past an hour
  endpoints: (builder) => ({
    getProducts: builder.query({
      query: () => "product",
      providesTags: ['Product']
    }),
    getProductById: builder.query({
      query: (productId) => `product/${productId}`,
      providesTags: ['Product']
    }),
    addProductToCart: builder.mutation({
      query: (product) => ({
        url: "/cart",
        method: "POST",
        body: product,
      }),
      invalidatesTags: ['Product']
    }),
  }),
});

export const {
  useGetProductsQuery,
  useGetProductByIdQuery,
  useAddProductToCartMutation,
} = productAPI;
