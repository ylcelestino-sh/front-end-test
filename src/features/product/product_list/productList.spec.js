import React from "react";
import {
  fireEvent,
  render,
} from "../../../utils/testUtils";
import ProductCard from "./ProductCard";
import * as router from "react-router";
import fetchMock from "jest-fetch-mock";
import Products from "./Product";
import { act } from "react-dom/test-utils";
import { products } from "../../../services/testData";

const navigate = jest.fn();

beforeEach(() => {
  jest.spyOn(router, "useNavigate").mockImplementation(() => navigate);
  fetchMock.resetMocks();
});

describe("Product list Tests", () => {
  test("Show Oops no data found when elementsSearched it's empty", async () => {
    const { getByTestId } = await act(async () => render(<Products />));
    expect(getByTestId("noDataFound")).toBeInTheDocument();
  });

  test("productCard show correctly and and navigate to /product:id on click", () => {
    const product = {
      brand: "Acer",
      id: "ZmGrkLRPXOTpxsU4jjAcv",
      imgUrl:
        "https://front-test-api.herokuapp.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg",
      model: "Iconia Talk S",
      price: "170",
    };
    const { getByTestId } = render(
      <ProductCard
        brand={product.brand}
        image={product.imgUrl}
        model={product.model}
        price={product.price}
        id={product.id}
      />
    );
    const productCard = getByTestId("productCard");
    expect(productCard).toMatchSnapshot();

    fireEvent.click(getByTestId("productCardBox"));

    expect(navigate).toHaveBeenCalledWith(`/product/${product.id}`);
  });

  // test("Show Spinner when data is fetching", async () => {
  //   fetchMock.mockResponseOnce(JSON.stringify(products));
  //   const storeRef = setupApiStore(productAPI);
  //   render(<Products />, { wrapper: storeRef.wrapper });
  //   await waitFor(() =>
  //     expect(screen.getByTestId("isLoading")).toBeInTheDocument()
  //   );
  // });

  test("show products when elementsSearched have data", async () => {
    fetchMock.mockResponseOnce(JSON.stringify(products));
    const { getByTestId } = await act(async () => render(<Products />));
    expect(getByTestId("productsData")).toBeInTheDocument();
  });
});
