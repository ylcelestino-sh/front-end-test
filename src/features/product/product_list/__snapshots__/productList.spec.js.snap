// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product list Tests productCard show correctly and and navigate to /product:id on click 1`] = `
<div
  class="css-1pxb9vs"
  data-testid="productCard"
>
  <button
    class="css-1rhsp8h"
    data-testid="productCardBox"
    role="group"
  >
    <div
      class="css-1th8op1"
    >
      <img
        alt="Iconia Talk S"
        class="chakra-image__placeholder css-zv8j3m"
        data-testid="lazyImage"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
      />
    </div>
    <div
      class="chakra-stack css-1fa334y"
    >
      <h1
        class="chakra-heading css-18j379d"
      >
        Iconia Talk S
      </h1>
      <h2
        class="chakra-heading css-18j379d"
      >
        $
        170
      </h2>
      <div
        class="chakra-stack css-n21gh5"
      >
        <p
          class="chakra-text css-m9w7kq"
        >
          brand: 
          Acer
        </p>
      </div>
    </div>
  </button>
</div>
`;
